programs/server/Server.cpp

```diff
@@ -1662,8 +1662,17 @@ try
         config().getString("path", DBMS_DEFAULT_PATH),
         std::move(main_config_zk_node_cache),
         main_config_zk_changed_event,
-        [&](ConfigurationPtr config, bool initial_loading)
+        [&, config_file = config().getString("config-file", "config.xml")](ConfigurationPtr config, bool initial_loading)
         {
+            if (!initial_loading)
+            {
+                /// Add back "config-file" key which is absent in the reloaded config.
+                config->setString("config-file", config_file);
+
+                /// Apply config updates in global context.
+                global_context->setConfig(config);
```

Comment:
FWIW @tavplubix shows some concerns about this in the past - https://github.com/ClickHouse/ClickHouse/pull/34574/files#r805753489

Reply:
cc: @amosbird @antaljanosbenjamin 

Also, `setConfig` replaces the `config` object (probably destroying the old one) while `getConfigRef()` returns a reference to this object. Should we revert?

Reply:
I'll try to address mentioned issues in #71452

Reply:
@tavplubix @azat Thanks for catching these issues, I completely missed these problems and trusted our CI, which I shouldn't have.